### PR TITLE
Config file support

### LINKED ISSUE #2558: Ability to pass a configuration file to Meilisearch
Discussed with @gmourier and following this information: [https://github.com/meilisearch/product/discussions/85#discussioncomment-2864656](https://github.com/meilisearch/product/discussions/85#discussioncomment-2864656)

We want Meilisearch to accept a configuration file that would be an alternative to the [command-line options](https://docs.meilisearch.com/learn/configuration/instance_options.html#command-line-options-and-flags) you can set when launching Meilisearch, for instance

```bash
meilisearch --master-key=XXX
```

# Requirements

### Format

The file format should be `TOML`.

### Priority order

Configuration file < [Env variables](https://docs.meilisearch.com/learn/configuration/instance_options.html#environment-variables) < [command-line options](https://docs.meilisearch.com/learn/configuration/instance_options.html#command-line-options-and-flags)

It means the options in the configuration file can be overwritten by the environment variable set (if they exist) and that can be themself overwritten by the command-line options (if they exist)

Example

With this config file

```bash
master_key = "thisisasuper1337masterkey@@@"
db_path = "./filepath1"
http_address = "127.0.0.1:7701"
no_anayltics = true
```

and with the following variables and command-line options:

```bash
export MEILI_DB_PATH=./filepath2
export MEILI_HTTP_ADDR=127.0.0.1:7702

meilisearch --master-key=abcd1234 --db-path=./filepath3
```

The final values kept by Meilisearch will be

- No analytics thanks to `no_analytics = true` set in the config file
- The db path is `filepath3` because `--db-path=./filepath3` has been set as command line options. `./filepath1` and `./filepath2` are ignored.
- The HTTP address is `127.0.0.1:7702` because set as an environment variable. The value in the config file (`127.0.0.1:7701`) is ignored.
- The master key is `abcd1234` set in the command-line option. The value in the config file is ignored ( `thisisasuper1337masterkey@@@` ).

### Config file content

<s>The config file will have sections gathering the same kind of variable.

Example:

```bash
master_key = "thisisasuper1337masterkey@@@"
no_analytics = true

[dump]
path = "dumps"

[snapshot]
enabled = true

[ssl]
require_auth = true
```
</s>

See [comments below](https://github.com/meilisearch/meilisearch/issues/2558#issuecomment-1239681961), sections are not something we want.

What we would rather: a default config file containing
- fake sections separated by comments
- a comment header redirecting to the documentation for more information

@meilisearch/docs-team this file could be easily downloaded 

```bash
curl https://raw.githubusercontent.com/meilisearch/meilisearch/main/config.toml > config.toml
```

### Config file command line

The default config file name is `config.toml` and will be retrieved by Meilisearch by default in the current working directory

You can pass the config file as command-line option

```bash
meilisearch --config-file-path="./config.toml"
```

So the default value of `config-file-path` is `./config.toml`.

This means by default, Meilisearch will try to read `./config.toml`.

If no `./config.toml` file is found and no `--config-file-path` option is passed, Meilisearch does not throw any error.
If `--config-file-path` is passed and is associated to an existing file (or impossible to open), Meilisearch throw an error

### Display when launching Meilisearch

```
888b     d888          d8b 888 d8b                                            888
8888b   d8888          Y8P 888 Y8P                                            888
88888b.d88888              888                                                888
888Y88888P888  .d88b.  888 888 888 .d8888b   .d88b.   8888b.  888d888 .d8888b 88888b.
888 Y888P 888 d8P  Y8b 888 888 888 88K      d8P  Y8b     "88b 888P"  d88P"    888 "88b
888  Y8P  888 88888888 888 888 888 "Y8888b. 88888888 .d888888 888    888      888  888
888   "   888 Y8b.     888 888 888      X88 Y8b.     888  888 888    Y88b.    888  888
888       888  "Y8888  888 888 888  88888P'  "Y8888  "Y888888 888     "Y8888P 888  888

Config file path:       "./config.toml"
Database path:		"./data.ms"
```

If the users do not pass any config file and no `./config.toml` exists, we must display `none` instead of `"./config.toml"`

### Errors (bonus)

We need to parse and display clear errors message for the configuration scheme. e.g. [https://docs.rs/miette/latest/miette/](https://docs.rs/miette/latest/miette/)

# Implementations

We could probably start by using [the toml crate](https://lib.rs/crates/toml) and declare a struct with a simple derive on it. We could start the job in [the options.rs file](https://github.com/meilisearch/meilisearch/blob/main/meilisearch-lib/src/options.rs) or maybe create a new one specifically for this config file.

If you have any questions regarding the feature or the implementation, please let us know about this issue! Every feedback or questions are welcome!

# TODO

- [ ] Merge the first iteration of this issue (https://github.com/meilisearch/meilisearch/pull/2745)
- [ ] Improve the default config file (assigned: @curquiza)
- [ ] Fix the following sub-issues
  - [ ] https://github.com/meilisearch/meilisearch/issues/2800
  - [ ] https://github.com/meilisearch/meilisearch/issues/2801
- [ ] Create the specification

--- 

## For Meili people

A spec will be open once we have implemented it. Some changes can still be applied, we need to see the final implementation to create the spec.
Since we let this issue (so far) to the community, we don't know yet in which release this new feature will be available (at minimum v0.29.0, not before).

Ping @meilisearch/cloud-team @meilisearch/docs-team who can be concerned by this issue.

### PR DESCRIPTION
**Pull request by mlemesle**
# Pull Request

## What does this PR do?
Fixes #2558

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

**Comment by curquiza** [comment]
@gmourier and @maryamsulemani97 I let you also review the categories in the default config file: I indeed think it's too bad the `master-key` and other useful commands like `db-path` and `http-address` are at the end of the file.

File: config.toml
Comment (maryamsulemani97) [inline config.toml:77]:
```suggestion
# Read the private key from KEYFILE.  This should be an RSA private key or PKCS8-encoded
```

**Review by maryamsulemani97 — CHANGES_REQUESTED** [review]
Hi @mlemesle, great work! I made a few suggestions to the `config.toml` file.
- The file is missing the disable analytics option
- I think it would be better if you started with the variable name and then the explanation(see suggestion on lines 7&8), it would make it easier to read
- I agree with @curquiza's comment on the order, I think you could go with the order used in the [docs](https://docs.meilisearch.com/learn/configuration/instance_options.html#all-instance-options)

File: meilisearch-http/src/option.rs
Comment (curquiza) [inline meilisearch-http/src/option.rs:261]:
```suggestion
                Err(err) => eprintln!("Can't read {:?} : {}", config_file_path, err),
```

**Review by curquiza — CHANGES_REQUESTED** [review]
About the config file sections, I'm not sure every option has to be in a section. For example, you can write the most important at the top without any section and then, add the other sections (dumps, snapshot...)

Also, a problem I encountered when trying your PR: Meilisearch does not read my config file. It runs without issue, but my variable in the config files are not taken into account by doing
```
cargo run --release -- --config-file-path="./config.toml"
```

and with this in the `config.toml`

```
db-path = "./lol.ms"
```

**Comment by mlemesle** [comment]
> Thank you again @mlemesle for the PR!!
> 
> About the config file sections, I'm not sure every option has to be in a section. For example, you can write the most important at the top without any section and then, add the other sections (dumps, snapshot...) I agree with @maryamsulemani97 to follow the documentation order, if possible!
> 
> Also, a problem I encountered when trying your PR: Meilisearch does not read my config file. It runs without issue, but my variable in the config files are not taken into account by doing
> 
> ```
> cargo run --release -- --config-file-path="./config.toml"
> ```
> 
> and with this in the `config.toml`
> 
> ```
> db-path = "./lol.ms"
> ```

Hey @curquiza, regarding your issue you’re writing the key using kebab-case but the example in the issue was in snake_case. I used snake_case, that’s why your value isn’t used.
If needs be, I can change it to kebab-case.
Please let me know 🙂

**Comment by curquiza** [comment]
Hello @mlemesle, thank you I better understand now!
I think the error handler of this flag should be improved then haha
I would rather review once all the requested changes have been applied, and the kebab-case issue as well, if you agree. This way we avoid multiple reviews 😄

File: meilisearch-http/src/option.rs
Comment (Kerollmops) [inline meilisearch-http/src/option.rs:270]: To make sure that we don't forget any struct field I would prefer that we destructure the `Opt` and list exactly *all of the fields*:

```rust
let Opt { db_path, env: _ } = self;
```

File: meilisearch-lib/src/options.rs
Comment (Kerollmops) [inline meilisearch-lib/src/options.rs:15]:
```suggestion
```

File: meilisearch-lib/src/options.rs
Comment (Kerollmops) [inline meilisearch-lib/src/options.rs:64]: Same as above, use the destructuring feature here.

File: meilisearch-lib/src/options.rs
Comment (Kerollmops) [inline meilisearch-lib/src/options.rs:109]: Same as above.

**Review by Kerollmops — CHANGES_REQUESTED** [review]
Hey, I really appreciate the work you've done. I am just a little worried that the whole set of variables in the default config file is all commented, it is hard to read and find the content. I don't know what we can do about that.

File: config.toml
Comment (irevoire) [inline config.toml:16]:
```suggestion
# master_key = "MASTER_KEY"
```

The config file is in snake case right?

**Review by irevoire — CHANGES_REQUESTED** [review]
almost there

File: config.toml
Comment (Kerollmops) [inline config.toml:73]: Would it be possible to check that it works with either a string like that and also a native number (without quotes)?

**Review by Kerollmops — CHANGES_REQUESTED** [review]
Could you please uncomment every option that has a default value that is static and not computed at runtime? It would be much better that way.

File: config.toml
Comment (mlemesle) [reply config.toml:73]: It works with native numbers but I think it’s less readable.
I can either use "100 GiB" or 100000000000, both works fine.

What’s your preference ?

File: config.toml
Comment (maryamsulemani97) [inline config.toml:93]:
```suggestion
# process if a database already exists or if no snapshot exists at the given path. If this
```

File: config.toml
Comment (maryamsulemani97) [inline config.toml:54]: Could you please remove this sentence? We don't have any recommendations as it depends on the usage

**Review by curquiza — CHANGES_REQUESTED** [review]
Hello again!

I talked with the core team and @maryamsulemani97: no need to apply any change on the default config file anymore. We are sorry not to have provided any clear information about it, and so we will open another PR to improve the tiny details we want to change.
Thanks again for your involvement even in this non-tech part, it was really appreciated! 😄 

Before merging, regarding the product, there is still the error message I would like to change, making the error more explicit rather the `Error: No such file or directory (os error 2)`

I suggest
"Error: impossible to open or read the XXX configuration file." if possible

Thanks again for your PR! ❤️

**Comment by mlemesle** [comment]
> Hello again!
> 
> I talked with the core team and @maryamsulemani97: no need to apply any change on the default config file anymore. We are sorry not to have provided any clear information about it, and so we will open another PR to improve the tiny details we want to change. Thanks again for your involvement even in this non-tech part, it was really appreciated! 😄
> 
> Before merging, regarding the product, there is still the error message I would like to change, making the error more explicit rather the `Error: No such file or directory (os error 2)`
> 
> I suggest "Error: impossible to open or read the XXX configuration file." if possible
> 
> Thanks again for your PR! ❤️

Hey @curquiza, 

No problem at all ! Regarding the error message, if I understand correctly you want this message to be printed only when a specific file is used by the user who will pass `--config-file-path` flag ? No need to print a message if the default file can’t be read for some reason ?

**Comment by curquiza** [comment]
> No problem at all ! Regarding the error message, if I understand correctly you want this message to be printed only when a specific file is used by the user who will pass --config-file-path flag ? No need to print a message if the default file can’t be read for some reason ?

If possible I would like for both
- when the default cannot be read
- if the passed one cannot be read

Let me know if it's not possible

**Comment by mlemesle** [comment]
> > No problem at all ! Regarding the error message, if I understand correctly you want this message to be printed only when a specific file is used by the user who will pass --config-file-path flag ? No need to print a message if the default file can’t be read for some reason ?
> 
> If possible I would like for both
> 
> * when the default cannot be read
> * if the passed one cannot be read
> 
> Let me know if it's not possible

@curquiza Should this be a blocking condition ? 
I saw a comment saying that we should display none if the default file can’t be read ?
So is there a difference between those two cases ?

Is that’s possible, can you tell me what should be the behavior in those cases ?
When the file is specified by the user should we stop if we can’t read or open it ? If not, where should be located the error message ?
When the file used is the default one, should we block if we can’t read it ? 

I’m kinda lost between all the messages tbh, can you clarify this for me please ? Sorry for the inconvenience

**Comment by curquiza** [comment]
Oh yes, sorry you're right, I was confused btw these behaviors!
The error message should only happen for config files we cannot open when the user passes the `--config-file-path`
Sorry again for the confusion!

**Comment by mlemesle** [comment]
No problem !
It's clear, I'll look into it.
Thanks for your patience!! 🙂

**Comment by mlemesle** [comment]
Hey @curquiza, 

Hope you’re doing fine ! I took your remarks into consideration, it allowed me to improve error management a bit more, thanks !!

As usual, feel free to reach out if needed 🙂

**Review by curquiza — APPROVED** [review]
Looks good to me! 😄 

Thank you very much @mlemesle and sorry again for the confusions!

We don't want to merge it
- before hacktoberfest has started
- at least one developer AND @gmourier have approved

**Comment by mlemesle** [comment]
No problem @curquiza, a pleasure to contribute !

That’s good for me, thanks !

**Review by gmourier — APPROVED** [review]
Thanks a lot @mlemesle! That's a big contribution to Meilisearch. 💪 

Sorry again for the back and forth; we discovered important aspects as we went along. We must define clear expectations before opening contributions to ease everyone's work.

**Review by irevoire — APPROVED** [review]
In terms of feature AND codewise it looks good to me, thanks for your contribution :tada:

**Review by Kerollmops — CHANGES_REQUESTED** [review]
Apart from the comment I added, I find the code very good!
Thank you very much for your contribution 🎉

File: meilisearch-lib/src/options.rs
Comment (curquiza) [reply meilisearch-lib/src/options.rs:25]: "It is recommended to use something like 80%-90% of the available memory, no more." -> this is not something we recommend, it's not even in the docs
Maryam asked to remove it here, in the config file, so I suppose @mlemesle wanted to be consistent -> https://github.com/meilisearch/meilisearch/pull/2745#discussion_r976455040

In any case, I opened this issue to improve the comments later, maybe we could avoid blocking the PR for the not-exact comments? 😊 
https://github.com/meilisearch/meilisearch/issues/2810

**Review by Kerollmops — APPROVED** [review]
Thank you for your PR, we will work on the comment details on our own! 🎉 🚀

**Comment by mlemesle** [comment]
Thanks everyone for letting me contribute to Meilisearch, it was a pleasure ! :rocket:

**Comment by curquiza** [comment]
bors merge

**Comment by mlemesle** [comment]
> Time has come!
> 
> 
> 
> bors merge
> 
> 
> 
> Thanks again @mlemesle!!! ❤️ 
> 
> 
> 
> 
> 
> If you are participating in [Hacktoberfest](https://blog.meilisearch.com/contribute-hacktoberfest-2022/), and you would like to receive a small gift from Meilisearch too, don't forget to fill [this form](https://meilisearch.typeform.com/to/idt5qugY).

Thanks @curquiza !
Was a pleasure or contribute.
Sadly, since the PR has been created before the opening of Hacktoberfest, it seems it doesn't count...

Guess I'll have to contribute more to Meilisearch 👀

**Comment by curquiza** [comment]
Oh nooooo!
But it definitely count for us, please subscribe to the form 😄

**Pull request merged**
